src/ripple/app/tx/impl/XChainBridge.cpp

```diff
@@ -0,0 +1,1627 @@
+
+            // let txn succeed if error distributing rewards (other than
+            // inability to pay)
+        }
+
+        if (distributed > rewardPool)
+            return tecINTERNAL;
+    }
+
+    return thTer;
+}
+
+/** Get signers list corresponding to the account that owns the bridge
+
+    @param view View to read the signer's list from.
+    @param sleB Sle of the bridge.
+    @param j Log
+
+    @return map of the signer's list (AccountIDs and weights), the quorum, and
+            error code
+*/
+std::tuple<std::unordered_map<AccountID, std::uint32_t>, std::uint32_t, TER>
+getSignersListAndQuorum(ApplyView& view, SLE const& sleB, beast::Journal j)
+{
+    std::unordered_map<AccountID, std::uint32_t> r;
+    std::uint32_t q = std::numeric_limits<std::uint32_t>::max();
+
+    auto const sleS = view.read(keylet::signers(sleB[sfAccount]));
+    if (!sleS)
+        return {r, q, tecXCHAIN_NO_SIGNERS_LIST};
+    q = (*sleS)[sfSignerQuorum];
+
+    auto const accountSigners = SignerEntries::deserialize(*sleS, j, "ledger");
+
+    if (!accountSigners)
+    {
+        return {r, q, tecINTERNAL};
+    }
+
+    for (auto const& as : *accountSigners)
+    {
+        r[as.account] = as.weight;
+    }
+
+    return {std::move(r), q, tesSUCCESS};
+};
+}  // namespace
+//------------------------------------------------------------------------------
+
+NotTEC
+BridgeCreate::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureXChainBridge))
+        return temDISABLED;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    if (ctx.tx.getFlags() & tfUniversalMask)
+        return temINVALID_FLAG;
+
+    auto const account = ctx.tx[sfAccount];
+    auto const reward = ctx.tx[sfSignatureReward];
+    auto const minAccountCreate = ctx.tx[~sfMinAccountCreateAmount];
+    auto const bridge = ctx.tx[sfXChainBridge];
+    if (bridge.lockingChainDoor() == bridge.issuingChainDoor())
+    {
+        return temEQUAL_DOOR_ACCOUNTS;
+    }
+
+    if (bridge.lockingChainDoor() != account &&
+        bridge.issuingChainDoor() != account)
+    {
+        return temSIDECHAIN_NONDOOR_OWNER;
+    }
+
+    if (isXRP(bridge.lockingChainIssue()) != isXRP(bridge.issuingChainIssue()))
+    {
+        // Because ious and xrp have different numeric ranges, both the src and
+        // dst issues must be both XRP or both IOU.
+        return temSIDECHAIN_BAD_ISSUES;
+    }
+
+    if (!isXRP(reward) || reward.signum() <= 0)
+    {
+        return temXCHAIN_BRIDGE_BAD_REWARD_AMOUNT;
+    }
+
+    if (minAccountCreate &&
+        (!isXRP(*minAccountCreate) || minAccountCreate->signum() <= 0))
+    {
+        return temXCHAIN_BRIDGE_BAD_MIN_ACCOUNT_CREATE_AMOUNT;
+    }
+
+    if (isXRP(bridge.issuingChainIssue()))
+    {
+        // Issuing account must be the root account for XRP (which presumably
+        // owns all the XRP). This is done so the issuing account can't "run
+        // out" of wrapped tokens.
```

Comment:
Couldn't someone construct a sidechain where they move all the XRP to a different account? i.e. the testnet has [this account](https://testnet.xrpl.org/accounts/rPT1Sjq2YGrBMTttX4GZHjKu9dyfzbpAYe) as the faucet account, which contains all the spare XRP

Reply:
Could we instead do a preclaim check to make sure that the bridge has almost all of the XRP? Not sure where the threshold should be, but it likely shouldn't be _all_ the XRP in case they've already done some setup work on the sidechain.

Reply:
For XRP to XRP bridges it's hard to see why it wouldn't always be the genesis account. I think we should keep the constraint, and if someone comes up with an important use-case where it's necessary for a different account to hold all the XRP we can loosen it later.